hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/test/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/fair/TestFSSchedulerNode.java

```diff
@@ -0,0 +1,376 @@
+    schedulerNode.releaseContainer(containers.get(0).getContainerId(), true);
+    assertEquals("Everything should have been released",
+        Resources.none(), schedulerNode.getAllocatedResource());
+
+    // Check that we are error prone
+    schedulerNode.releaseContainer(containers.get(0).getContainerId(), true);
+    finalValidation(schedulerNode);
+  }
+
+  /**
+   * Allocate and release three containers with launch.
+   */
+  @Test
+  public void testMultipleAllocations() {
+    RMNode node = createNode();
+    FSSchedulerNode schedulerNode = new FSSchedulerNode(node, false);
+
+    createDefaultContainer();
+    createDefaultContainer();
+    createDefaultContainer();
+    assertEquals("Nothing should have been allocated, yet",
+        Resources.none(), schedulerNode.getAllocatedResource());
+    schedulerNode.allocateContainer(containers.get(0));
+    schedulerNode.containerStarted(containers.get(0).getContainerId());
+    schedulerNode.allocateContainer(containers.get(1));
+    schedulerNode.containerStarted(containers.get(1).getContainerId());
+    schedulerNode.allocateContainer(containers.get(2));
+    assertEquals("Container should be allocated",
+        Resources.multiply(containers.get(0).getContainer().getResource(), 3.0),
+        schedulerNode.getAllocatedResource());
+    schedulerNode.releaseContainer(containers.get(1).getContainerId(), true);
+    schedulerNode.releaseContainer(containers.get(2).getContainerId(), true);
+    schedulerNode.releaseContainer(containers.get(0).getContainerId(), true);
+    finalValidation(schedulerNode);
+  }
+
+  /**
+   * Allocate and release a single container.
+   */
+  @Test
+  public void testSimplePreemption() {
+    RMNode node = createNode();
+    FSSchedulerNode schedulerNode = new FSSchedulerNode(node, false);
+
+    // Launch containers and saturate the cluster
+    saturateCluster(schedulerNode);
+    assertEquals("Container should be allocated",
+        Resources.multiply(containers.get(0).getContainer().getResource(),
+            containerNum),
+        schedulerNode.getAllocatedResource());
+
+    // Request preemption
+    FSAppAttempt starvingApp = createStarvingApp(schedulerNode,
+        Resource.newInstance(1024, 1));
+    schedulerNode.addContainersForPreemption(
+        Collections.singletonList(containers.get(0)), starvingApp);
+    assertEquals(
+        "No resource amount should be reserved for preemptees",
+        containers.get(0).getAllocatedResource(),
+        schedulerNode.getTotalReserved());
+
+    // Preemption occurs
+    schedulerNode.releaseContainer(containers.get(0).getContainerId(), true);
+    allocateContainers(schedulerNode);
+    assertEquals("Container should be allocated",
+        schedulerNode.getTotalResource(),
+        schedulerNode.getAllocatedResource());
+
+    // Release all containers
+    for (int i = 1; i < containerNum; ++i) {
+      schedulerNode.releaseContainer(containers.get(i).getContainerId(), true);
```

Comment:
It is not 0 because I release the rest of the containers that were not preempted. That is, I would like to see the actual schedulerNode.releaseContainer call next to the preemption in the same function to see what is actually happening. Another function would hide the ability to see all releases together.